csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -114,18 +114,27 @@ std::vector<TensorView*> getOutputsWithoutMesh(Expr* expr) {
 // Returns the set of parallel types not seen on the loop domain of the given
 // tvs and hence, can be propagated.
 std::unordered_set<ParallelType> getParallelTypesToPropagate(
-    std::vector<TensorView*> tvs) {
-  // Get the set of parallel types seen on the loop domain of the given tvs.
+    std::vector<TensorView*> tvs,
+    bool is_input = false) {
+  std::unordered_set<ParallelType> all_parallel_types;
+  if (is_input) {
+    all_parallel_types = {ParallelType::Stream};
+  } else {
+    all_parallel_types = deviceAndStreamParallelTypes();
+  }
+
+  // Collect any DID or stream parallel types seen on the loop domain of the
+  // given tvs. For DID, only non-reduction dimensions are considered.
   std::unordered_set<ParallelType> existing_parallel_types;
   for (auto tv : tvs) {
     for (auto id : tv->getLoopDomain()) {
-      if (!id->isReduction() && id->isDeviceDim()) {
+      if ((id->isDeviceDim() && !id->isReduction()) || id->isStream()) {
```

Comment:
Why is Stream handled differently from DIDs? What's wrong with e.g. backprop'ing
```
  m
 / \
s*
```
back to
```
  m, r(s*)
```
?

Reply:
I didn't understand your example -- is `r(s*)` stream-parallel?
The condition is:
1. Propagate a DID dimension to a tensorview if it is not already present on a non-reduction dimension. This produces the reduce-scatter pattern.
2. If the reduction dimension is stream parallelized, we can still not have another stream parallel id unlike DID parallelization. For e.g.: `i0, Stream(r{s}), r{i1/s}` cannot be stream parallelized on `i0` when propagating. Am I missing any case here?

Reply:
Resolved offline.